tests/test_es_privilege_check.py

```diff
@@ -38,19 +37,73 @@ async def test_bad_connection(hass: HomeAssistantType, es_aioclient_mock: Aiohtt
 
     await gateway.async_stop_gateway()
 
+
 @pytest.mark.asyncio
-async def test_successful_privilege_check(hass: HomeAssistantType, es_aioclient_mock: AiohttpClientMocker):
+async def test_successful_modern_privilege_check(
+    hass: HomeAssistantType, es_aioclient_mock: AiohttpClientMocker
+):
     """Test successful privilege check."""
 
-    es_url = "http://test_successful_privilege_check:9200"
+    es_url = "http://test_successful_modern_privilege_check:9200"
+    mock_es_initialization(es_aioclient_mock, es_url)
+
+    config = build_full_config(
+        {"url": es_url, CONF_INDEX_MODE: "datastream", CONF_USERNAME: "test"}
+    )
+
+    gateway = ElasticsearchGateway(config)
+    await gateway.async_init()
+
+    instance = ESPrivilegeCheck(gateway)
+    result = await instance.check_privileges(config)
+    assert result.has_all_requested
+
+    result = await instance.enforce_privileges(config)
+    assert result is None
+
+    await gateway.async_stop_gateway()
+
+
+@pytest.mark.asyncio
+async def test_successful_modern_privilege_check_missing_index_privilege(
+    hass: HomeAssistantType, es_aioclient_mock: AiohttpClientMocker
+):
+    """Test successful privilege check with missing index privileges."""
+
+    es_url = "http://test_successful_modern_privilege_check:9200"
     mock_es_initialization(
-        es_aioclient_mock,
-        es_url
+        es_aioclient_mock, es_url, mock_index_authorization_error=True
     )
 
-    config = build_full_config({
-        "url": es_url
-    })
+    config = build_full_config(
+        {"url": es_url, CONF_INDEX_MODE: "datastream", CONF_USERNAME: "test"}
+    )
+
+    gateway = ElasticsearchGateway(config)
+    await gateway.async_init()
+
+    instance = ESPrivilegeCheck(gateway)
+    result = await instance.check_privileges(config)
+    assert result.has_all_requested is False
+    assert result.missing_cluster_privileges == []
+    assert result.missing_index_privileges == {"hass-events*": ["index"]}
```

Comment:
This outcome is a bit misleading. We're using the existing `mock_index_authorization_error` flag above, which simulates missing privileges for the legacy index format, as opposed to the new datastream name.